libvast/src/format/csv.cpp

```diff
@@ -129,4 +140,271 @@ const char* writer::name() const {
+        return data{pattern{std::move(x)}};
+      };
+    } else if constexpr (std::is_same_v<T, set_type>) {
+      auto set_insert = [](std::vector<Attribute>&& xs) {
+        return set(std::make_move_iterator(xs.begin()),
+                   std::make_move_iterator(xs.end()));
+      };
+      return ('{'_p >> (caf::visit(*this, t.value_type) % set_separator_)
+              >> '}'_p)
+               ->*set_insert;
+    } else if constexpr (std::is_same_v<T, vector_type>) {
+      auto vector_insert = [](std::vector<Attribute>&& xs) { return xs; };
+      return ('[' >> (caf::visit(*this, t.value_type) % set_separator_) >> ']')
+               ->*vector_insert;
+    } else if constexpr (std::is_same_v<T, map_type>) {
+      auto map_insert = [](std::vector<std::tuple<Attribute, Attribute>>&& xs) {
+        auto to_pair = [](auto&& tuple) {
+          return std::make_pair(std::get<0>(tuple), std::get<1>(tuple));
+        };
+        map m;
+        for (auto& x : xs)
+          m.insert(to_pair(std::move(x)));
+        return m;
+      };
+      auto kvp = caf::visit(*this, t.key_type) >> '='
+                 >> caf::visit(*this, t.value_type);
+      return (('{'_p >> (kvp % set_separator_) >> '}')->*map_insert);
+    } else if constexpr (has_parser_v<type_to_data<T>>) {
+      using value_type = type_to_data<T>;
+      auto ws = ignore(*parsers::space);
+      return (ws >> make_parser<value_type>{} >> ws)->*[](value_type&& x) {
+        return x;
+      };
+    } else {
+      VAST_ERROR_ANON("csv parser builder faild to fetch a parser for type",
+                      caf::detail::pretty_type_name(typeid(T)));
+      return {};
+    }
+  }
+
+  std::string set_separator_;
+};
+
+template <class Iterator>
+struct csv_parser_factory {
+  using result_type = type_erased_parser<Iterator>;
+
+  csv_parser_factory(const std::string& set_separator,
+                     table_slice_builder_ptr bptr)
+    : set_separator_{set_separator}, bptr_{std::move(bptr)} {
+    // nop
+  }
+
+  struct identity {
+    template <class T>
+    constexpr T operator()(T&& t) const noexcept {
+      return std::forward<T>(t);
+    }
+  };
+
+  template <class T, class F>
+  struct add_t {
+    void operator()(caf::optional<T>&& x) const {
+      if (x) {
+        const auto tmp = f_(*x);
+        bptr_->add(make_data_view(tmp));
+      } else
+        bptr_->add(make_data_view(caf::none));
+    }
+    table_slice_builder_ptr bptr_;
+    F f_;
+  };
+
+  template <class T>
+  static add_t<T, identity> make_add_t(table_slice_builder_ptr bptr) {
+    return add_t<T, identity>{std::move(bptr), identity{}};
+  }
+
+  template <class T, class F>
+  static add_t<T, F> make_add_t(table_slice_builder_ptr bptr, F f) {
+    return add_t<T, F>{std::move(bptr), std::move(f)};
+  }
+
+  template <class T>
+  result_type operator()(const T& t) const {
+    if constexpr (std::is_same_v<T, string_type>) {
+      return (-+(parsers::any - set_separator_))
```

Comment:
This should probably be the standard separator, not the *set*-separator.

Reply:
Good catch.